docs/docs/use_cases/rag/code_understanding.mdx

```diff
@@ -0,0 +1,386 @@
+
+- We test max marginal relevance for retrieval
+- And 5 documents returned
+
+In this example we'll be using Supabase, however you can pick any vector store with MMR search you'd like from [our large list of integrations](/docs/modules/data_connection/vectorstores/integrations/).
+
+```typescript
+import { createClient } from "@supabase/supabase-js";
+import { OpenAIEmbeddings } from "langchain/embeddings/openai";
+import { SupabaseVectorStore } from "langchain/vectorstores/supabase";
+```
+
+```typescript
+const privateKey = process.env.SUPABASE_PRIVATE_KEY;
+if (!privateKey) throw new Error(`Expected env var SUPABASE_PRIVATE_KEY`);
+
+const url = process.env.SUPABASE_URL;
+if (!url) throw new Error(`Expected env var SUPABASE_URL`);
+
+const client = createClient(url, privateKey);
+```
+
+Once we've initialized our client we can pass it, along with some more options to the `.fromDocuments` method on `SupabaseVectorStore`.
+
+For more instructions on how to set up Supabase, see the [Supabase docs](/docs/modules/data_connection/vectorstores/integrations/supabase).
+
+```typescript
+const vectorStore = await SupabaseVectorStore.fromDocuments(
+  texts,
+  new OpenAIEmbeddings(),
+  {
+    client,
+    tableName: "documents",
+    queryName: "match_documents",
+  }
+);
+```
+
+```typescript
+const retriever = vectorStore.asRetriever({
+  searchType: "mmr", // Use max marginal relevance search
+  searchKwargs: { fetchK: 5 },
+});
+```
+
+### Chat
+
+We'll setup our model and memory system just as we'd do for any other chatbot application.
+
+```typescript
+import { ChatOpenAI } from "langchain/chat_models/openai";
+```
+
+```typescript
+const model = new ChatOpenAI({ modelName: "gpt-4" });
+```
+
+We're going to use `BufferMemory` as our memory chain. All this will do is take in inputs/outputs from the LLM and store them in memory.
+
+```typescript
+import { BufferMemory } from "langchain/memory";
+```
+
+```typescript
+const memory = new BufferMemory({
+  returnMessages: true, // Return stored messages as instances of `BaseMessage`
+  memoryKey: "chat_history", // This must match up with our prompt template input variable.
+});
+```
+
+Now we can construct our main sequence of chains. We're going to be building `ConversationalRetrievalChain` using Expression Language.
+
+```typescript
+import { ChatPromptTemplate, MessagesPlaceholder } from "langchain/prompts";
+import { RunnableBranch, RunnableSequence } from "langchain/schema/runnable";
+import { formatDocumentsAsString } from "langchain/util/document";
+import { BaseMessage } from "langchain/schema";
+import { StringOutputParser } from "langchain/schema/output_parser";
+```
+
+First we construct the combine documents chain. This chain takes in the question, chat history (from memory) and context (relevant documents). It then passes all three into the prompt which answers the users question.
+
+```typescript
+const combineDocumentsPrompt = ChatPromptTemplate.fromMessages([
+  [
+    "ai",
```

Comment:
This is the common paradigm:

```js
    const chatPrompt = ChatPromptTemplate.fromMessages([
      SystemMessagePromptTemplate.fromTemplate(
        "You are a helpful assistant who must always respond like a pirate"
      ),
      new MessagesPlaceholder("history"),
      HumanMessagePromptTemplate.fromTemplate("{input}"),
    ]);
```

This prompting pattern will work for some models but not others (e.g. Anthropic)